### PR TITLE
ruleset: add support for more complex skip rulesets

### LINKED ISSUE #1177: Add support SKIP rules and rulesets on WAF (ruleset)
### Current Terraform and Cloudflare provider version

```
➜ terraform -v
Terraform v1.0.5
on darwin_amd64
+ provider registry.terraform.io/cloudflare/cloudflare v2.26.1
+ provider registry.terraform.io/hashicorp/random v2.3.1
```

### Description

Not all skip functionalities are supported yet, like `rules` and `rulesets` in `action_parameters` for a rules with an action of skip.

#### Skip

For `rules` and `rulesets` the structure described in the api doc doesn't seem to be implemented in the `cloudflare-go` [here](https://github.com/cloudflare/cloudflare-go/blob/master/rulesets.go#L170) neither in the `terraform-provider-cloudflare`.

### Use cases

The use case is a configuration of payload logging/skip directly via a terraform configuration as the same as the UI/API.

### Potential Terraform configuration

```hcl
# Zone-level WAF Managed Ruleset
resource "cloudflare_ruleset" "zone_level_managed_waf" {
  zone_id     = var.cloudflare_zone_id
  name        = "managed WAF"
  description = "managed WAF ruleset description"
  kind        = "zone"
  phase       = "http_request_firewall_managed"

  rules {
    action = "skip"
    action_parameters {
        rulesets = ["efb7b8c949ac4650a09736fc376e9aee"]
    }
    expression = "(cf.zone.name eq \"domain.xyz\" and http.request.uri.query contains \"skip=rulesets\")"
    description = "skip Cloudflare Manage ruleset"
    enabled = true 
  }

  rules {
    action = "skip"
    action_parameters {
       # efb7b8c949ac4650a09736fc376e9aee is the ruleset ID of the Cloudflare Managed rules
        rules = {
          "efb7b8c949ac4650a09736fc376e9aee" = [
                        "5de7edfa648c4d6891dc3e7f84534ffa",
                        "e3a567afc347477d9702d9047e97d760"
                    ]
        }
    }
    expression = "(cf.zone.name eq \"domain.xyz\" and http.request.uri.query contains \"skip=rules\")"
    description = "skip Wordpress rule and sQLi rule"
    enabled = true 
  }

  rules {
    action = "execute"
    action_parameters {
        id = "efb7b8c949ac4650a09736fc376e9aee"
        version = "latest"
        overrides {
          rules {
                  id = "5de7edfa648c4d6891dc3e7f84534ffa"
                  action = "block"
                  enabled = true
          }
          rules {
            id = "75a0060762034a6cb663fd51a02344cb"
            action = "log"
            enabled = true
          }
          categories {
            category = "wordpress"
            action = "js_challenge"
            enabled = true
          }
        }
    }
    expression = "true"
    description = "Execute Cloudflare Managed Ruleset on my zone-level phase entry point ruleset"
    enabled = true
  }
}
```

### References

- Skip `rules` documentation [here](https://developers.cloudflare.com/waf/managed-rulesets/waf-exceptions/define-api#skip-one-or-more-rules-of-waf-managed-rulesets)
- Skip `rulesets` documentation [here](https://developers.cloudflare.com/waf/managed-rulesets/waf-exceptions/define-api#skip-one-or-more-waf-managed-rulesets)

### PR DESCRIPTION
**Pull request by jacobbednarz**
Depends on cloudflare/cloudflare-go#708

Closes #1177

### CODE REVIEW

File: website/docs/r/ruleset.html.markdown [hunk 1]
```diff
@@ -237,7 +237,9 @@ The following arguments are supported:
 * `increment` - (Optional)
 * `overrides` - (Optional) List of override configurations to apply to the ruleset (refer to the [nested schema](#nestedblock--action-parameters-overrides)).
 * `products` - (Optional) Products to target with the actions. Valid values are `"bic"`, `"hot"`, `"ratelimit"`, `"securityLevel"`, `"uablock"`, `"waf"` or `"zonelockdown"`.
-* `ruleset` - (Optional) Which ruleset to target. Valid value is `"current"`.
+* `ruleset` - (Optional) Which ruleset ID to target.
```
Comment (jacobbednarz) [inline website/docs/r/ruleset.html.markdown:240]: cc @pedrosousa

File: cloudflare/resource_cloudflare_ruleset_test.go [hunk 2]
```diff
@@ -1329,6 +1379,68 @@ func testAccCheckCloudflareRulesetRateLimit(rnd, name, zoneID, zoneName string)
...
+      description = "skip Cloudflare Manage ruleset"
+      enabled = true
+    }
+
+    rules {
+      action = "skip"
+      action_parameters {
+        # efb7b8c949ac4650a09736fc376e9aee is the ruleset ID of the Cloudflare Managed rules
+        rules = {
+          "efb7b8c949ac4650a09736fc376e9aee" = "5de7edfa648c4d6891dc3e7f84534ffa,e3a567afc347477d9702d9047e97d760"
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_ruleset_test.go:1410]: i don't love that `TypeMap` only allows simple types as values however this is the next best thing to having it as a list of strings and just perform the translation in the provider.

**Comment by jacobbednarz** [comment]
with cloudflare-go shimmed in, all looking good

```
TF_ACC=1 go test $(go list ./...) -v -run "^TestAccCloudflareRuleset_" -count 1 -parallel 1 -timeout 120m -parallel 1
?   	github.com/cloudflare/terraform-provider-cloudflare	[no test files]
=== RUN   TestAccCloudflareRuleset_WAFBasic
=== PAUSE TestAccCloudflareRuleset_WAFBasic
=== RUN   TestAccCloudflareRuleset_WAFManagedRuleset
=== PAUSE TestAccCloudflareRuleset_WAFManagedRuleset
=== RUN   TestAccCloudflareRuleset_WAFManagedRulesetOWASP
=== PAUSE TestAccCloudflareRuleset_WAFManagedRulesetOWASP
=== RUN   TestAccCloudflareRuleset_WAFManagedRulesetOWASPBlockXSSWithAnomalyOver60
=== PAUSE TestAccCloudflareRuleset_WAFManagedRulesetOWASPBlockXSSWithAnomalyOver60
=== RUN   TestAccCloudflareRuleset_WAFManagedRulesetOWASPOnlyPL1
=== PAUSE TestAccCloudflareRuleset_WAFManagedRulesetOWASPOnlyPL1
=== RUN   TestAccCloudflareRuleset_WAFManagedRulesetDeployMultiple
=== PAUSE TestAccCloudflareRuleset_WAFManagedRulesetDeployMultiple
=== RUN   TestAccCloudflareRuleset_WAFManagedRulesetDeployMultipleWithSkip
=== PAUSE TestAccCloudflareRuleset_WAFManagedRulesetDeployMultipleWithSkip
=== RUN   TestAccCloudflareRuleset_WAFManagedRulesetDeployMultipleWithTopSkipAndLastSkip
=== PAUSE TestAccCloudflareRuleset_WAFManagedRulesetDeployMultipleWithTopSkipAndLastSkip
=== RUN   TestAccCloudflareRuleset_WAFManagedRulesetWithCategoryBasedOverrides
=== PAUSE TestAccCloudflareRuleset_WAFManagedRulesetWithCategoryBasedOverrides
=== RUN   TestAccCloudflareRuleset_WAFManagedRulesetWithIDBasedOverrides
=== PAUSE TestAccCloudflareRuleset_WAFManagedRulesetWithIDBasedOverrides
=== RUN   TestAccCloudflareRuleset_MagicTransitUpdateWithHigherPriority
    provider_test.go:138: Skipping acceptance test as 0da42c8d2132a9ddaf714f9e7c920711 is not configured for Magic Transit
--- SKIP: TestAccCloudflareRuleset_MagicTransitUpdateWithHigherPriority (0.00s)
=== RUN   TestAccCloudflareRuleset_WAFManagedRulesetWithPayloadLogging
=== PAUSE TestAccCloudflareRuleset_WAFManagedRulesetWithPayloadLogging
=== RUN   TestAccCloudflareRuleset_RateLimit
=== PAUSE TestAccCloudflareRuleset_RateLimit
=== RUN   TestAccCloudflareRuleset_TransformationRuleURIPath
=== PAUSE TestAccCloudflareRuleset_TransformationRuleURIPath
=== RUN   TestAccCloudflareRuleset_TransformationRuleURIQuery
=== PAUSE TestAccCloudflareRuleset_TransformationRuleURIQuery
=== RUN   TestAccCloudflareRuleset_TransformationRuleHeaders
=== PAUSE TestAccCloudflareRuleset_TransformationRuleHeaders
=== RUN   TestAccCloudflareRuleset_ActionParametersMultipleSkips
=== PAUSE TestAccCloudflareRuleset_ActionParametersMultipleSkips
=== CONT  TestAccCloudflareRuleset_WAFBasic
--- PASS: TestAccCloudflareRuleset_WAFBasic (3.80s)
=== CONT  TestAccCloudflareRuleset_WAFManagedRulesetWithIDBasedOverrides
--- PASS: TestAccCloudflareRuleset_WAFManagedRulesetWithIDBasedOverrides (3.70s)
=== CONT  TestAccCloudflareRuleset_WAFManagedRulesetDeployMultiple
--- PASS: TestAccCloudflareRuleset_WAFManagedRulesetDeployMultiple (3.46s)
=== CONT  TestAccCloudflareRuleset_WAFManagedRulesetWithCategoryBasedOverrides
--- PASS: TestAccCloudflareRuleset_WAFManagedRulesetWithCategoryBasedOverrides (2.96s)
=== CONT  TestAccCloudflareRuleset_WAFManagedRulesetDeployMultipleWithTopSkipAndLastSkip
--- PASS: TestAccCloudflareRuleset_WAFManagedRulesetDeployMultipleWithTopSkipAndLastSkip (4.14s)
=== CONT  TestAccCloudflareRuleset_WAFManagedRulesetDeployMultipleWithSkip
--- PASS: TestAccCloudflareRuleset_WAFManagedRulesetDeployMultipleWithSkip (3.54s)
=== CONT  TestAccCloudflareRuleset_TransformationRuleURIQuery
--- PASS: TestAccCloudflareRuleset_TransformationRuleURIQuery (2.91s)
=== CONT  TestAccCloudflareRuleset_ActionParametersMultipleSkips
--- PASS: TestAccCloudflareRuleset_ActionParametersMultipleSkips (3.34s)
=== CONT  TestAccCloudflareRuleset_TransformationRuleHeaders
--- PASS: TestAccCloudflareRuleset_TransformationRuleHeaders (2.95s)
=== CONT  TestAccCloudflareRuleset_WAFManagedRulesetOWASPBlockXSSWithAnomalyOver60
--- PASS: TestAccCloudflareRuleset_WAFManagedRulesetOWASPBlockXSSWithAnomalyOver60 (3.54s)
=== CONT  TestAccCloudflareRuleset_WAFManagedRulesetOWASPOnlyPL1
--- PASS: TestAccCloudflareRuleset_WAFManagedRulesetOWASPOnlyPL1 (3.17s)
=== CONT  TestAccCloudflareRuleset_RateLimit
--- PASS: TestAccCloudflareRuleset_RateLimit (3.17s)
=== CONT  TestAccCloudflareRuleset_TransformationRuleURIPath
--- PASS: TestAccCloudflareRuleset_TransformationRuleURIPath (3.19s)
=== CONT  TestAccCloudflareRuleset_WAFManagedRulesetWithPayloadLogging
--- PASS: TestAccCloudflareRuleset_WAFManagedRulesetWithPayloadLogging (2.86s)
=== CONT  TestAccCloudflareRuleset_WAFManagedRulesetOWASP
--- PASS: TestAccCloudflareRuleset_WAFManagedRulesetOWASP (2.86s)
=== CONT  TestAccCloudflareRuleset_WAFManagedRuleset
--- PASS: TestAccCloudflareRuleset_WAFManagedRuleset (3.24s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/cloudflare	53.512s
?   	github.com/cloudflare/terraform-provider-cloudflare/version	[no test files]
```

File: website/docs/r/ruleset.html.markdown (hunk 1 above)
Comment (jacobbednarz) [reply website/docs/r/ruleset.html.markdown:240]: i'm going to land this PR (for the sake of needing to cut a breaking change sometime this week) however feel free to raise a PR if this isn't suitable in your own time and we can get this addressed.

Comment (pedrosousa) [reply website/docs/r/ruleset.html.markdown:240]: Sorry, for some reason I missed this one... I'll create a separate PR. 👍🏼

**Pull request merged**